libraries/ESP8266mDNS/examples/LEAmDNS/mDNS_Clock_v2/mDNS_Clock_v2.ino

```diff
@@ -0,0 +1,282 @@
+/*
+  ESP8266 mDNS responder clock
+
+  This example demonstrates two features of the LEA clsLEAMDNSHost:
+  1. The host and service domain negotiation process that ensures
+     the uniqueness of the finally choosen host and service domain name.
+  2. The dynamic MDNS service TXT feature
+
+  A 'clock' service in announced via the MDNS responder and the current
+  time is set as a TXT item (eg. 'curtime=Mon Oct 15 19:54:35 2018').
+  The time value is updated every second!
+
+  The ESP is initially announced to clients as 'esp8266.local', if this host domain
+  is already used in the local network, another host domain is negociated. Keep an
+  eye to the serial output to learn the final host domain for the clock service.
+  The service itself is is announced as 'host domain'._espclk._tcp.local.
+  As the service uses port 80, a very simple HTTP server is installed also to deliver
+  a small web page containing a greeting and the current time (not updated).
+  The web server code is taken nearly 1:1 from the 'mDNS_Web_Server.ino' example.
+  Point your browser to 'host domain'.local to see this web page.
+
+  Instructions:
+  - Update WiFi SSID and password as necessary.
+  - Flash the sketch to the ESP8266 board
+  - Install host software:
+    - For Linux, install Avahi (http://avahi.org/).
+    - For Windows, install Bonjour (http://www.apple.com/support/bonjour/).
+    - For Mac OSX and iOS support is built in through Bonjour already.
+  - Use a MDNS/Bonjour browser like 'Discovery' to find the clock service in your local
+    network and see the current time updates.
+
+*/
+
+
+#include <ESP8266WiFi.h>
+#include <WiFiClient.h>
+#include <ESP8266WebServer.h>
+#include <LwipIntf.h>
+#include <time.h>
+
+// uses API MDNSApiVersion::LEAv2
+
+/*
+   Include the clsLEAMDNSHost (the library needs also to be included)
+   As LEA clsLEAMDNSHost is experimantal in the ESP8266 environment currently, the
+   legacy clsLEAMDNSHost is defaulted in the include file.
+   There are two ways to access LEA clsLEAMDNSHost:
+   1. Prepend every declaration and call to global declarations or functions with the namespace, like:
+      'LEAmDNS::clsLEAMDNSHost::hMDNSService  hMDNSService;'
+      This way is used in the example. But be careful, if the namespace declaration is missing
+      somewhere, the call might go to the legacy implementation...
+   2. Open 'ESP8266mDNS.h' and set LEAmDNS to default.
+
+*/
+#define NO_GLOBAL_MDNS // our MDNS is defined below
```

Comment:
> This doesn't really avoid the MDNS global instance, it just disables the extern statement in the .h . That means that the global instance in the .cpp may still exist.

I renamed the instance in this example and just verified with `readelf -a`: the default global instance `MDNS` is not there.

> In addition, if any other .cpp includes the header, that TU will have access to the global instance, while this .ino will be working with the instance declared here.

This is user responsability to include MDNS in one place or to use this `NO_GLOBAL_MDNS` define everytime.

>  I'm thinking of the Arduino ota thing.

This is true, but this is beyond this pull request that did not introduce `NO_GLOBAL_MDNS` but just makes use of it.
An issue should be opened about that. Maybe arduino OTA should be given a parameter instead of declaring/using its own mDNS instance.

>  Such macros need to be defined globally in order to assure correctness.

I don't currently know any *easy* and *in arduino spirit* way of declaring a global define, when using the Arduino IDE.

Reply:
> This is user responsibility to include MDNS in one place or to use this NO_GLOBAL_MDNS define everytime

You can't push this responsibility on the user. I'm sure that even I wouldn't get it right, there are just too many ways to get it wrong even with just user code using this. With ArduinoOTA and 3rd party libs in general that could make use of MDNS, the chances of getting it right are almost nil. We will continue to get issue reports about MDNS not working correctly, we'll have to spend our effort on this, the symptoms will be along the lines of "it fails for the user but works for us", and in the end no resolution will be reached.

In addition, this is an example, and users tend to base their apps on examples. At the very least this requires a Big Phat Comment above explaining that this way of doing things does not cover use of the global MDNS instance by libraries, whether user or 3rd party, and that the "correct" way of doing it is to add a global define in the build parameters. How to do that can be explained in detail in readthedocs, and a link provided here.

Having said that, I do have an idea to at least detect the problem case: make all instances of the MDNS class aware of each other, and check each other for errors. E.g.: have a static member list with ptrs to all instances. When a class constructs, it adds itself to the list, when it destructs it removes itself. A debug message could be printed out that shows the number of items on the list. If we allow only one instance of MDNS at any given time, the list can be simplified to a simple flag. If not, when an instance calls begin(), check the list for any other instance that would cause conflict, and error out or crater if that is the case.

Reply:
> You can't push this responsibility on the user

This is not specific to this PR.

I agree when it comes to ArduinoOTA. That one should be changed to force user to give an MDNS instance, be it the default global `MDNS` one. 
Otherwise, that's the current model also used for r.i.p. SPIFFS, LittleFS, SDFS, EEPROM, SPI, ...
For all these classes, including mDNS, the `NO_GLOBAL_INSTANCES` / `NO_GLOBAL_XXX` has always applied: either define this globally (how to do that ? there's the unfriendly `platform.local.txt` - at least not really documented in our doc) or use `#define NO_GLOBAL_...` before `#include <>`.

It'd be surely interesting to have a mechanism allowing to detect whether the default global instance is used in conjunction with another user instance (when `#include <mDNS>` is used in two different files with only one `NO_GLOBAL_`).